core/node/registries/river_registry_contract.go

```diff
@@ -393,13 +431,155 @@ func (c *RiverRegistryContract) ForAllStreams(
 			if err != nil {
 				return err
 			}
+			totalStreams++
 			if !cb(makeGetStreamResult(streamId, &stream.Stream)) {
 				return nil
 			}
 		}
 	}
 
-	log.Info("RiverRegistryContract: GetPaginatedStreams completed", "elapsed", time.Since(startTime))
+	elapsed := time.Since(startTime)
+	log.Info(
+		"RiverRegistryContract: GetPaginatedStreams completed",
+		"elapsed",
+		elapsed,
+		"streamsPerSecond",
+		float64(totalStreams)/elapsed.Seconds(),
+	)
+
+	return nil
+}
+
+func (c *RiverRegistryContract) forAllStreamsParallel(
+	ctx context.Context,
+	blockNum crypto.BlockNumber,
+	cb func(*GetStreamResult) bool,
+) error {
+	log := dlog.FromCtx(ctx)
+	ctx, cancelCtx := context.WithCancel(ctx)
+	defer cancelCtx()
+
+	numWorkers := c.Settings.ParallelReaders
+	if numWorkers <= 1 {
+		numWorkers = 8
+	}
+
+	pageSize := int64(c.Settings.PageSize)
+	if pageSize <= 0 {
+		pageSize = 5000
+	}
+
+	progressReportInterval := c.Settings.ProgressReportInterval
+	if progressReportInterval <= 0 {
+		progressReportInterval = 10 * time.Second
+	}
+
+	numStreamsBigInt, err := c.StreamRegistry.GetStreamCount(c.callOptsWithBlockNum(ctx, blockNum))
+	if err != nil {
+		return WrapRiverError(Err_CANNOT_CALL_CONTRACT, err).Func("ForAllStreams")
+	}
+	numStreams := numStreamsBigInt.Int64()
+
+	if numStreams <= 0 {
+		log.Info("RiverRegistryContract: GetPaginatedStreams no streams found", "blockNum", blockNum)
+		return nil
+	}
+
+	log.Info(
+		"RiverRegistryContract: GetPaginatedStreams starting parallel read",
+		"numStreams",
+		numStreams,
+		"RiverRegistry.PageSize",
+		pageSize,
+		"RiverRegistry.ParallelReaders",
+		numWorkers,
+		"blockNum",
+		blockNum,
+	)
+
+	chResults := make(chan []river.StreamWithId, numWorkers)
+	chErrors := make(chan error, numWorkers)
+
+	pool := workerpool.New(numWorkers)
+
+	startTime := time.Now()
+	lastReport := time.Now()
+	var taskCounter atomic.Int64
+	for i := int64(0); i < numStreams; i += pageSize {
+		taskCounter.Add(1)
+		pool.Submit(func() {
+			streams, _, err := c.callGetPaginatedStreamsWithBackoff(ctx, blockNum, i, i+pageSize)
+			if ctx.Err() == nil {
+				if err == nil {
+					chResults <- streams
```

Comment:
On both `chResults` and `chErrors` you can use a write + read select to prevent a deadlock entirely. Something like this:

```
select {
case chResults <- streams:
   break 
case <-ctx.Done():
   break
}
```
This takes away the need to check if the ctx was cancelled with an error.